src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -458,125 +400,71 @@ private void StartSendBlob(byte[] incoming, int count, AsyncProtocolRequest asyn
-                {
-                    InnerStream.Write(message.Payload, 0, message.Size);
-                }
-                else
-                {
-                    asyncRequest.AsyncState = message;
-                    Task t = InnerStream.WriteAsync(message.Payload, 0, message.Size, asyncRequest.CancellationToken);
-                    if (t.IsCompleted)
-                    {
-                        t.GetAwaiter().GetResult();
-                    }
-                    else
-                    {
-                        IAsyncResult ar = TaskToApm.Begin(t, s_writeCallback, asyncRequest);
-                        if (!ar.CompletedSynchronously)
-                        {
-#if DEBUG
-                            asyncRequest._DebugAsyncChain = ar;
-#endif
-                            return;
-                        }
-                        TaskToApm.End(ar);
-                    }
-                }
+                InnerStream.Write(message.Payload, 0, message.Size);
             }
 
-            CheckCompletionBeforeNextReceive(message, asyncRequest);
+            CheckCompletionBeforeNextReceive(message);
         }
 
         //
         // This will check and logically complete / fail the auth handshake.
         //
-        private void CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
+        private void CheckCompletionBeforeNextReceive(ProtocolToken message)
         {
             if (message.Failed)
             {
-                StartSendAuthResetSignal(null, asyncRequest, ExceptionDispatchInfo.Capture(new AuthenticationException(SR.net_auth_SSPI, message.GetException())));
+                StartSendAuthResetSignal(null, ExceptionDispatchInfo.Capture(new AuthenticationException(SR.net_auth_SSPI, message.GetException())));
                 return;
             }
-            else if (message.Done && !_pendingReHandshake)
+            else if (message.Done)
             {
                 ProtocolToken alertToken = null;
 
                 if (!CompleteHandshake(ref alertToken))
                 {
-                    StartSendAuthResetSignal(alertToken, asyncRequest, ExceptionDispatchInfo.Capture(new AuthenticationException(SR.net_ssl_io_cert_validation, null)));
+                    StartSendAuthResetSignal(alertToken, ExceptionDispatchInfo.Capture(new AuthenticationException(SR.net_ssl_io_cert_validation, null)));
                     return;
                 }
 
                 // Release waiting IO if any. Presumably it should not throw.
                 // Otherwise application may get not expected type of the exception.
-                FinishHandshake(null, asyncRequest);
+                FinishHandshake(null);
                 return;
             }
 
-            StartReceiveBlob(message.Payload, asyncRequest);
+            StartReceiveBlob(message.Payload);
         }
 
         //
         // Server side starts here, but client also loops through this method.
```

Comment:
It seems like these methods were split up into pieces in order to facilitate async operations.  Can they be recombined, using looping rather than recursion, to simplify it all?

Reply:
yes, I think so. I originally tried to preserve names and structure but as I removed more code consolidation makes more sense.